agent/util-scripts/pbench-tool-meister-start

```diff
@@ -1,33 +1,47 @@
 #!/usr/bin/env python3
 # -*- mode: python -*-
 
-"""pbench-tool-meister-start
+"""pbench-tool-meister-start - start the execution of the Tool Meister
+sub-system.
 
-Responsible for orchestrating the Redis Server, Tool Data Sink (TDS), and Tool
-Meisters (TMs):
+If the user does not specify the `--redis-server` command line parameter (or
+the `PBENCH_REDIS_SERVER` environment variable is not present, or is an empty
+string), tool-meister-start is responsible for orchestrating the creation of a
+Redis Server, Tool Data Sink (TDS), and Tool Meisters (TMs).  If the user does
+specify `--redis-server` (or environment variable), tool-meister-start assumes
+the user has already instantiated a TDS and appropriate TMs, and begins
```

Comment:
Isn't there a `--redis-server` and a `--tool-data-sink` option though? This seems to imply that using `--redis-server` means that redis, tds, and tms have all already been setup by the user. If I just use the --redis-server option, is it then assuming the tds is running locally (`--redis-server` doesn't require `--tool-data-sink` according to usage)? I feel like if I was using only that option, I would expect it to only assume I set up a redis server myself. Also why does --tool-data-sink then require `--redis-server`, would it not go off that same assumption that it used when only `--redis-server` was used (for the tds)? 
The point is, I think with the newly added level of abstraction in this file, it could probably support more than "all or nothing" for pre-set-up components. Is there a reason why we can't have `--redis-server` only mean that the user already has a redis server at `<hostname>:<port>`, have `--tool-data-sink` only mean that the user has a tds already at `<hostname>:<port>`, and maybe a `--tm premade` option to indicate that the user already has tms running on the registered hostnames? They would then be able to mix and match according to their needs (maybe they have an existing redis server they want to use and need us to make tds/tms, or maybe they are using their own tds/tm containers but want us to provide a redis-server).

Reply:
Added flexibility means both added logic and added testing to handle all the cases.

At this time, it seems best to support only one or the other.

Reply:
When I originally started to read this comment, I leaped to a different conclusion almost immediately from the very first question:  Why have command line options (and associated environment variables) and then have a second set of environment variables to change how the orchestration is performed?

As a result, I am thinking of changing the interface to be that there is a new switch, `--orchestrate` which accepts the keyword options: `create` and `existing`.  When `--orchestrate=create` is used, it directs `pbench-tool-meister-start` to create its own instance of a Redis server, a Tool Data Sink, and create both the local and remote Tool Meister instances.  When `--orchestrate=existing` is used, it directs `pbench-tool-meister-start` to expect the caller to have performed the orchestration, and then to expect to be told how to contact the Redis server and Tool Data Sink via the `--redis-server` and `--tool-data-sink` options.  When `--orchestrate` is not present, it would default to `create` to allow for the current behavior of `agent/bench-scripts` to work without further changes.

For example, to support:
 * The current behavior of our `agent/bench-scripts`, no changes
 * Using IP/port combinations different from the defaults, the user would define `PBENCH_REDIS_SERVER` and `PBENCH_TOOL_DATA_SINK` to have the required values (changing the `--redis-server` and `--tool-data-sink` default values given)
 * A Kubernetes environment, one would specify `--redis-server=<...>` & `--tool-data-sink=<...>` (or as environment variables) without `--orchestrate`, and `pbench-tool-meister-start` would be contact them as directed

This would allow for one way to specify the IP/port combinations (command line options and environment variables), and then direct the behavior requested explicitly.  Further, the keyword options make it easier to extend the behavior in the future if desired.

Thoughts?

Reply:
I'm amused:  last week, I wrote up a lengthy reply to Mustafa's comment, but then it got attached to the rest of my review (thanks, Github...), and, by the time I was ready to post it, you (@portante) had already replied, so I just deleted it.  I was going to suggest (among other things) an option not unlike what you outline here.

Since we're intending only two modes of support, it does seem like we should have a binary set of options.  The interface should make it clear to each of the user and the tool which mode we're supposed to be in, and then the parsing and validation should be able to be handled in an orderly fashion.

A switch like `--orchestrate` seems good, but I would avoid providing keyword values to it, unless you seriously think we'll be adding additional values.  I suggest we stick with a binary choice, like `--orchestrate` and `--no-orchestrate` (particularly if our command parser of choice supports negatable options).  Obviously, omitting the option would yield the appropriate default (that being, presumably, having Pbench create the processes).

That said, you seem to be looking to make the `--orchestrate` switch basically orthogonal to the other switches.  (That seems good....)  So, if `--no-orchestrate` is specified, then the agent will _not_ create the processes; otherwise (e.g., if the switch is omitted), it _will_ create them all.  And, then, `--redis-server` and `--tool-data-sync` (etc.) are used to specify where the processes can be contacted, _regardless_ of who created them.

Sounds good.

The only testable constraint (that comes to my mind) is, if `--no-orchestrate` is specified, then both `--redis-server` and `--tool-data-sync` _must_ be specified; otherwise, we have applicable defaults for all cases (I think).

Reply:
> The only testable constraint (that comes to my mind) is, if `--no-orchestrate` is specified, then both `--redis-server` and `--tool-data-sync` _must_ be specified; otherwise, we have applicable defaults for all cases (I think).

Agreed.

Reply:
I kinda like the `--orchestrate` binary. I'm not really crazy about `--no-orchestrate` but the twist here is that the default probably needs to be the historic "orchestrate" behavior of start/stop, so we're likely stuck with it. In any case, I like that better than the `--orchestrate=<keyword>` alternative.

Then again, this still implies "all or nothing" and even though you're not supporting anything else right now, we may well end up moving towards a "split orchestration" model where someone will expect to point us at an existing Redis server but let us orchestrate the data sync and tool meisters because that's purely internal Pbench-ese.

This makes me think we might be better off without `--orchestrate` at all, and simply rely on `--redis-server` to override the internal Redis orchestration and `--tool-data-sync` to override that orchestration. Right now we could require either no overrides or all overrides because you don't internally support splits, but that could easily be relaxed later. Why do we need a second related option? The default becomes "full orchestration because you haven't specified existing endpoints", which is historic behavior.

Reply:
What we currently have does that, but it uses 2 different sets of environment variables to make that happen, and requires that both `--tool-data-sink` and `--redis-server` be used.

The new proposal allows us to operate with no additional options in the default case, and use only one pair of environment variables by adding the `--orchestrate=<keyword>` switch (which avoids adding a second `--no-orchestrate`.

I'll add a commit that implements the new method so that we can decide.

Reply:
If we have appropriate command line parser support, then `--orchestrate` and `--no-orchestrate` are opposite senses of _the same option_ (e.g., if both forms appear on the same command line, it will either be automagically caught as an error or the second occurrence will override the first -- either way, from an implementation perspective, it's a __single__ option).

Reply:
Apparently `ArgumentParser` doesn't handle negation of keyword options, and doing so is therefore something of a hack. Another option would be simply `--orchestrate=[yes|no]` ... or we could make it a choice keyword of `["none", "all", "redis", "tms", "tds"]` ;-)

Reply:
Yeah, if we don't have support for negatable command line options, then the road gets bumpier.  The problem is that we want the _absence_ of the option to be the same as the __positive__ assertion, which means that we need the _presence_ of the option to be the __negative__ assertion, which means we need to name the option `--no-orchestrate` or some positive synonym for it (and lets forget about implementing an `--orchestrate` option).

I don't think that `--orchestrate=[yes|no]` is significantly better than `--orchestrate=[create|existing]`.  (And, @dbutenhof , your more flexible suggestion left out `"tms-and-tds"`, et al.  ;-) )

I'll go look at Peter's new  commit when its ready, but perhaps something like `--external_orchestration` would work (if it is present, we don't create the processes; if it's absent, we do).

Reply:
> I don't think that `--orchestrate=[yes|no]` is significantly better than `--orchestrate=[create|existing]`. (And, @dbutenhof , your more flexible suggestion left out `"tms-and-tds"`, et al. ;-) )

Actually, my unstated assumption was that `ArgumentParser` could support a _list_ of those keyword values; although after glancing at the documentation I'm not certain that it can. But, if it could, the default would be `all` (traditional behavior) while a fully externally-orchestrated environment would be `none`. And if we aren't immediately going to support mix-and-match, then the other values are irrelevant for now.

Whether this is an improvement or just a side comment... well, yeah, that's not so clear.

Reply:
It support that with `choices`, see the latest commit, which I believe makes this a bit more compact.